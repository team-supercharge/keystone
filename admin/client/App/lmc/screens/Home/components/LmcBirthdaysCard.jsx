import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import moment from 'moment';
import LmcResidentListItem from '../../../components/LmcResidentListItem.jsx';


class LmcBirthdaysCard extends Component {
    constructor (props) {
        super(props);
        this.getBirthdays = this.getBirthdays.bind(this);
    }

    renderBirthdays (birthdays) {
        let data = birthdays.map(({ id, name, picture, nextBirthday, age }) => {
            let reminderText;
            if (nextBirthday === 0) {
                reminderText = 'today!';
            } else if (nextBirthday === 1) {
                reminderText = 'tomorrow';
            } else if (nextBirthday < 6) {
                reminderText = `this coming ${moment().add(nextBirthday, 'days').format('dddd')}`;
            } else {
                reminderText = `on ${moment().add(nextBirthday, 'days').format('dddd Do MMM')}`;
            };

            return {
                name: `${age} years old ${reminderText}`,
                subheading: `${name}`,
                id,
                picture,
            };
        });

        return (
            <div>
                <h2 className="lmc-card-title">
                    { TITLE }
                </h2>
                <div className="lmc-card">
                    <div className="lmc-card-body">
                        { data.map(resident => (
                            <LmcResidentListItem
                                key={resident.id}
                                data={resident}
                                link={d => `${Keystone.adminPath}/residents/${d}`} />
                        )) }
                    </div>
                </div>
            </div>
        );
    }

    getBirthdays (residents) {
        return _.chain(residents)
            .filter({ status: 'active' })
            .cloneDeep()
            .forEach(res => {
                const { age, nextBirthday } = this.daysUntil(res.dateOfBirth);
                res.age = age;
                res.nextBirthday = nextBirthday;
            })
            .sortBy('nextBirthday')
            .filter(({ nextBirthday }) => nextBirthday < 120)
            .take(6)
            .value();
    }

    daysUntil (date) {
        const birthday = moment(date);
        const today = moment().format('YYYY-MM-DD');
        const age = moment(today).diff(birthday, 'years');
        const nextBirthday = moment(birthday).add(age, 'years');

        if (nextBirthday.isSame(today)) {
            return {
                nextBirthday: nextBirthday.diff(today, 'days'),
                age,
            };
        } else {
            /* add one more year in case the birthday has already passed
            to calculate date till next one. */
            return {
                nextBirthday: moment(birthday).add(age + 1, 'years').diff(today, 'days'),
                age: age + 1,
            };
        }
    }

    render () {
        const { residents } = this.props;
        const birthdays = this.getBirthdays(residents);
        return (
            <div>
                { birthdays && birthdays.length
                    ? this.renderBirthdays(birthdays)
                    : null }
            </div>
        );
    }
}

LmcBirthdaysCard.propTypes = {
    residents: PropTypes.array.isRequired,
};

const TITLE = 'Birthdays';

export default LmcBirthdaysCard;
