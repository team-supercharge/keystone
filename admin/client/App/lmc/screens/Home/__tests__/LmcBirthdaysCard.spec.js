import React from 'react'
import { shallow } from 'enzyme'
import LmcBirthdaysCard from '../components/LmcBirthdaysCard'
import MockDate from 'mockdate'

MockDate.set('1/1/2019')

describe('LmcBirthdaysCard', () => {
    let wrapper
    let residents
    let birthday

    beforeEach(() => {
        birthday = new Date('01-07-1940')
        residents = [
            {
                id: 1,
                dateOfBirth: birthday
            }
        ]
        wrapper = shallow((
            <LmcBirthdaysCard
                residents={residents}
            />
        ))
    })

    test('should render correct snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })

    test('should display correct upcoming birthday', () => {
        const birthdayAlert = wrapper.find('LmcResidentListItem')
        expect(birthdayAlert.props().data.name).toEqual('79 years old on Monday 7th Jan')
    })

    afterAll(() => {
        MockDate.reset()
    })
})