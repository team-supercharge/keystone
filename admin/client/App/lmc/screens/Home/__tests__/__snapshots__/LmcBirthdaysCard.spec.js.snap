// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LmcBirthdaysCard should render correct snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <h2
      className="lmc-card-title"
    >
      Birthdays
    </h2>
    <div
      className="lmc-card"
    >
      <div
        className="lmc-card-body"
      >
        <div>
          <LmcResidentListItem
            data={
              Object {
                "id": 1,
                "name": "79 years old on Monday 7th Jan",
                "picture": undefined,
                "subheading": "undefined",
              }
            }
            link={[Function]}
          />
        </div>
      </div>
    </div>
  </div>,
  "nodes": Array [
    <div>
      <h2
        className="lmc-card-title"
      >
        Birthdays
      </h2>
      <div
        className="lmc-card"
      >
        <div
          className="lmc-card-body"
        >
          <div>
            <LmcResidentListItem
              data={
                Object {
                  "id": 1,
                  "name": "79 years old on Monday 7th Jan",
                  "picture": undefined,
                  "subheading": "undefined",
                }
              }
              link={[Function]}
            />
          </div>
        </div>
      </div>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <LmcBirthdaysCard
        residents={
          Array [
            Object {
              "dateOfBirth": 1940-01-07T00:00:00.000Z,
              "id": 1,
              "status": "active",
            },
          ]
        }
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": LmcBirthdaysCard {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "getBirthdays": [Function],
        "props": Object {
          "residents": Array [
            Object {
              "dateOfBirth": 1940-01-07T00:00:00.000Z,
              "id": 1,
              "status": "active",
            },
          ],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <h2
            className="lmc-card-title"
          >
            Birthdays
          </h2>
          <div
            className="lmc-card"
          >
            <div
              className="lmc-card-body"
            >
              <div>
                <LmcResidentListItem
                  data={
                    Object {
                      "id": 1,
                      "name": "79 years old on Monday 7th Jan",
                      "picture": undefined,
                      "subheading": "undefined",
                    }
                  }
                  link={[Function]}
                />
              </div>
            </div>
          </div>
        </div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          <h2
            className="lmc-card-title"
          >
            Birthdays
          </h2>
          <div
            className="lmc-card"
          >
            <div
              className="lmc-card-body"
            >
              <div>
                <LmcResidentListItem
                  data={
                    Object {
                      "id": 1,
                      "name": "79 years old on Monday 7th Jan",
                      "picture": undefined,
                      "subheading": "undefined",
                    }
                  }
                  link={[Function]}
                />
              </div>
            </div>
          </div>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <LmcBirthdaysCard
    residents={
      Array [
        Object {
          "dateOfBirth": 1940-01-07T00:00:00.000Z,
          "id": 1,
          "status": "active",
        },
      ]
    }
  />,
}
`;
